pkg/authentication/scram/scram_credentials.go

```diff
@@ -0,0 +1,196 @@
+	if err != nil {
+		return nil, nil, fmt.Errorf("error generating stored key: %s", err)
+	}
+
+	serverKey, err = generateClientOrServerKey(hashConstructor, saltedPassword, serverKeyInput)
+	if err != nil {
+		return nil, nil, fmt.Errorf("error generating server key: %s", err)
+	}
+
+	return storedKey, serverKey, err
+}
+
+func generateB64EncodedSecrets(hashConstructor func() hash.Hash, password, b64EncodedSalt string, iterationCount int) (storedKey, serverKey string, err error) {
+	salt, err := base64.StdEncoding.DecodeString(b64EncodedSalt)
+	if err != nil {
+		return "", "", fmt.Errorf("error decoding salt: %s", err)
+	}
+
+	unencodedStoredKey, unencodedServerKey, err := generateSecrets(hashConstructor, password, salt, iterationCount)
+	if err != nil {
+		return "", "", fmt.Errorf("error generating secrets: %s", err)
+	}
+
+	storedKey = base64.StdEncoding.EncodeToString(unencodedStoredKey)
+	serverKey = base64.StdEncoding.EncodeToString(unencodedServerKey)
+	return storedKey, serverKey, nil
+}
+
+// password should be encrypted in the case of SCRAM-SHA-1 and unencrypted in the case of SCRAM-SHA-256
+func computeScramCredentials(hashConstructor func() hash.Hash, iterationCount int, base64EncodedSalt string, password string) (*automationconfig.ScramCreds, error) {
+	storedKey, serverKey, err := generateB64EncodedSecrets(hashConstructor, password, base64EncodedSalt, iterationCount)
+	if err != nil {
+		return nil, fmt.Errorf("error generating SCRAM-SHA keys: %s", err)
+	}
+
+	return &automationconfig.ScramCreds{IterationCount: iterationCount, Salt: base64EncodedSalt, StoredKey: storedKey, ServerKey: serverKey}, nil
+}
+
+// computeScram1AndScram256Credentials takes in a username and password, and generates SHA1 & SHA256 credentials
+// for that user. This should only be done if credentials do not already exist.
+func computeScram1AndScram256Credentials(resourceName, username, password string) (*automationconfig.ScramCreds, *automationconfig.ScramCreds, error) {
+	scram256Salt := getDeterministicSalt(resourceName, sha256.New)
+	scram1Salt := getDeterministicSalt(resourceName, sha1.New)
+
+	scram256Creds, err := computeCreds(username, password, scram256Salt, Sha256)
+	if err != nil {
+		return nil, nil, fmt.Errorf("error generating scramSha256 creds: %s", err)
+	}
+	scram1Creds, err := computeCreds(username, password, scram1Salt, Sha1)
+	if err != nil {
+		return nil, nil, fmt.Errorf("error generating scramSha1Creds: %s", err)
+	}
+	return scram1Creds, scram256Creds, nil
+}
+
+// getDeterministicSalt returns a deterministic salt based on the name of the resource.
+// the required number of characters will be taken based on the requirements for the SCRAM-SHA-1/MONGODB-CR algorithm
+func getDeterministicSalt(resourceName string, hashConstructor func() hash.Hash) []byte {
+	sha256bytes32 := sha256.Sum256([]byte(fmt.Sprintf("%s-mongodbresource", resourceName)))
+	return sha256bytes32[:hashConstructor().Size()-RFC5802MandatedSaltSize]
+}
```

Comment:
these 2 functions were added. We want to generate both Sha1 and Sha256 credentials as it bypasses a lot of issues with automation config validation.

A deterministic salt ensures that we only change the automation config on password changes.